File: src/main/java/binaryTree/Serialize.java
package binaryTree;

import java.io.*;

public interface Serialize {
    public void serialize(BinaryTree tree, String filename, String type);

    public BinaryTree deserialize(String filename);

    abstract class Abstract implements Serialize {
        @Override
        public void serialize(BinaryTree tree, String filename, String type) {
            try (PrintWriter writer = new PrintWriter(filename)) {
                writer.println(type);
                tree.forEachFromRoot(new ElementProcessor<UserType>() {
                    @Override
                    public void toDo(UserType v) {
                        writer.print(" ");
                        writer.print(v);
                    }
                });
            } catch (FileNotFoundException e) {
                e.printStackTrace();
            }
        }

        @Override
        public BinaryTree deserialize(String filename) {
            try (BufferedReader bufferedReader = new BufferedReader(new FileReader(filename))) {
                String type = bufferedReader.readLine();
                UserFactory userFactory = new UserFactory();
                if (!userFactory.getTypeNameList().contains(type)) {
                    throw new IllegalArgumentException("Wrong type");
                }

                String line;
                BinaryTree tree = new BinaryTree.Base();
                while ((line = bufferedReader.readLine()) != null) {
                    String[] items = line.split(" ");

                    for (String item : items) {
                        UserType builder = userFactory.getBuilderByName(type);
                        Object object = builder.parseValue(item);
                        if (object != null) {
                            tree.add((UserType) object);
                        }
                    }
                }
                return tree;
            } catch (IOException e) {
                e.printStackTrace();
            }
            return null;
        }

    }

    class Base extends Abstract {

    }
}

File: src/main/scala/binaryTreeScala/Serialize.scala
package binaryTreeScala

import java.io.{BufferedReader, FileNotFoundException, FileReader, IOException, PrintWriter}

trait Serialize {
  def serialize(tree: BinaryTree, filename: String, `type`: String): Unit

  def deserialize(filename: String): BinaryTree
}

object Serialize {
  abstract class Abstract extends Serialize {
    override def serialize(tree: BinaryTree, filename: String, `type`: String): Unit = {
      try {
        val writer = new PrintWriter(filename)
        writer.println(`type`)
        tree.forEachFromRoot(new ElementProcessor[UserType] {
          override def toDo(v: UserType): Unit = {
            writer.print(" ")
            writer.print(v)
          }
        })
        writer.close()
      } catch {
        case e: FileNotFoundException =>
          e.printStackTrace()
      }
    }

    override def deserialize(filename: String): BinaryTree = {
      try {
        val bufferedReader = new BufferedReader(new FileReader(filename))
        val `type` = bufferedReader.readLine()
        val userFactory = new UserFactory
        if (!userFactory.getTypeNameList.contains(`type`)) {
          throw new IllegalArgumentException("Wrong type")
        }

        var line: String = null
        val tree = new BinaryTree.Base
        while ( {
          line = bufferedReader.readLine();
          line
        } != null) {
          val items = line.split(" ")

          for (item <- items) {
            val builder = userFactory.getBuilderByName(`type`)
            val `object` = builder.parseValue(item)
            if (`object` != null) {
              tree.add(`object`.asInstanceOf[UserType])
            }
          }
        }
        tree
      } catch {
        case e: IOException =>
          e.printStackTrace()
          null
      }
    }
  }
  class Base extends Abstract {}
}



